cmd/multiclusterapp.go

```diff
@@ -0,0 +1,792 @@
+						Usage: "Show revisions available to rollback to",
+					},
+				},
+			},
+			cli.Command{
+				Name:      "upgrade",
+				Usage:     "Upgrade an app to a newer version",
+				Action:    multiClusterAppUpgrade,
+				ArgsUsage: "[APP_NAME/APP_ID VERSION]",
+				Flags: []cli.Flag{
+					cli.StringFlag{
+						Name:  "answers,a",
+						Usage: "Path to an answers file, the format of the file is a map with key:value. Supports JSON and YAML",
+					},
+					cli.StringFlag{
+						Name:  "values",
+						Usage: "Path to a helm values file.",
+					},
+					cli.StringSliceFlag{
+						Name:  "set",
+						Usage: "Set answers for the template, can be used multiple times. Example: --set foo=bar",
+					},
+					cli.BoolFlag{
+						Name:  "show-versions,v",
+						Usage: "Display versions available to upgrade to",
+					},
+					cli.StringSliceFlag{
+						Name: "target,t",
+						Usage: "Target project names/ids to upgrade. Specified targets on upgrade will override all " +
+							"the original targets. Leave it empty to keep current targets",
+					},
+				},
+			},
+			cli.Command{
+				Name:        "list-templates",
+				Aliases:     []string{"lt"},
+				Usage:       "List templates available for installation",
+				Description: "\nList all app templates in the current Rancher server",
+				ArgsUsage:   "None",
+				Action:      globalTemplateLs,
+				Flags: []cli.Flag{
+					formatFlag,
+					cli.StringFlag{
+						Name:  "catalog",
+						Usage: "Specify the catalog to list templates for",
+					},
+				},
+			},
+			cli.Command{
+				Name:        "show-template",
+				Aliases:     []string{"st"},
+				Usage:       "Show versions available to install for an app template",
+				Description: "\nShow all available versions of an app template",
+				ArgsUsage:   "[TEMPLATE_ID]",
+				Action:      templateShow,
+			},
+			cli.Command{
+				Name:      "show-app",
+				Aliases:   []string{"sa"},
+				Usage:     "Show an app's available versions and revisions",
+				ArgsUsage: "[APP_NAME/APP_ID]",
+				Action:    showMultiClusterApp,
+				Flags: []cli.Flag{
+					formatFlag,
+				},
+			},
+		},
+	}
+}
+
+func multiClusterAppLs(ctx *cli.Context) error {
+	c, err := GetClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	collection, err := c.ManagementClient.MultiClusterApp.List(defaultListOpts(ctx))
+	writer := NewTableWriter([][]string{
+		{"NAME", "App.Name"},
+		{"STATE", "App.State"},
+		{"VERSION", "Version"},
+		{"TARGET_PROJECTS", "Targets"},
+	}, ctx)
+
+	defer writer.Close()
+
+	templateVersionCache := map[string]*managementClient.TemplateVersion{}
+	clusterCache := map[string]*managementClient.Cluster{}
+	projectCache := map[string]*managementClient.Project{}
+	for _, item := range collection.Data {
+		version, err := getTemplateVersion(c.ManagementClient, templateVersionCache, item.TemplateVersionID)
+		if err != nil {
+			return err
+		}
+		targetNames, err := getReadableTargetNames(c, clusterCache, projectCache, item.Targets)
```

Comment:
If getReadableTargetNames returns an error lets just use the targets that already exist on the app vs returning the error.

Reply:
Updated to show ids when names are not applicable